compiler/luci/pass/src/helpers/HuffmanEncoder.h

```diff
@@ -0,0 +1,191 @@
+template <typename T> struct CompareNodes
+{
+  bool operator()(std::shared_ptr<Node<T>> l, std::shared_ptr<Node<T>> r)
+  {
+    return l->freq > r->freq;
+  }
+};
+
+template <typename T> class HuffmanEncoder
+{
+private:
+  std::unordered_map<T, std::string> _huffman_table;
+
+private:
+  std::shared_ptr<Node<T>> allocateNode(T data, unsigned int freq, std::shared_ptr<Node<T>> p_left,
+                                        std::shared_ptr<Node<T>> p_right)
+  {
+    std::shared_ptr<Node<T>> node = std::make_unique<Node<T>>();
+    node->data = data;
+    node->freq = freq;
+    node->p_left = p_left;
+    node->p_right = p_right;
+    return node;
+  }
+
+  std::unordered_map<T, unsigned int> calculateFrequencyMap(const std::vector<T> &input)
+  {
+    std::unordered_map<T, unsigned int> out_map;
+    for (auto &item : input)
+      out_map[item] = out_map.find(item) != out_map.end() ? out_map[item] + 1 : 1;
+    return out_map;
+  }
+
+  std::string exportHuffmanTreeToString(std::shared_ptr<Node<T>> node)
+  {
+    if (node == nullptr)
+      return "";
+    if (!node->p_left && !node->p_right)
+    {
+      return "0" + std::bitset<sizeof(T) * CHAR_BIT>(node->data).to_string();
+    }
+    std::string tmp = "1";
+    tmp += exportHuffmanTreeToString(node->p_left);
+    tmp += exportHuffmanTreeToString(node->p_right);
+    return tmp;
+  }
+
+  void buildHuffmanTable(std::shared_ptr<Node<T>> node, const std::string str = "")
+  {
+    if (node == nullptr)
+      return;
+
+    if (!node->p_left && !node->p_right)
+    {
+      _huffman_table[node->data] = str;
+    }
+
+    buildHuffmanTable(node->p_left, str + "0");
+    buildHuffmanTable(node->p_right, str + "1");
+  }
+
+  std::shared_ptr<Node<T>> buildHuffmanTree(const std::vector<T> &input)
+  {
+    auto freq_map = calculateFrequencyMap(input);
+
+    std::priority_queue<std::shared_ptr<Node<T>>, std::vector<std::shared_ptr<Node<T>>>,
+                        CompareNodes<T>>
+      pq;
+
+    for (auto &item : freq_map)
+    {
+      pq.push(allocateNode(item.first, item.second, nullptr, nullptr));
+    }
+    while (pq.size() != 1)
+    {
+      std::shared_ptr<Node<T>> left = pq.top();
+      pq.pop();
+      std::shared_ptr<Node<T>> right = pq.top();
+      pq.pop();
+
+      unsigned int sum = left->freq + right->freq;
+      pq.push(allocateNode(0, sum, left, right));
+    }
+
+    return pq.top();
+  }
+
+  struct EncodedTreeAndData
+  {
+    std::vector<bool> tree_vec{};
+    std::vector<bool> data_vec{};
+  };
+
+  std::vector<uint8_t> packEncodedDataToArray(const std::string &tree_str,
+                                              const std::string &encoded_data)
+  {
+    std::vector<uint8_t> arr;
+    const size_t kTreeSizeInBits = tree_str.size();
+    const size_t kDataSizeInBits = encoded_data.size();
+    for (size_t i = 0; i < sizeof(size_t); ++i)
+    {
+      arr.push_back(
+        *(static_cast<const uint8_t *>(static_cast<const void *>(&kTreeSizeInBits)) + i));
```

Comment:
can you plz split lines?

Reply:
@seanshpark Maybe think about better place for HuffmanEncoder.h/HuffmanDecoder.h library? It will be used in onert-micro and other inferences - so in current configuration luci dependency is required, can we place this on top level as separate dependency?

Reply:
> can we place this on top level as separate dependency?

Need to think about this. this maybe the first one, ... and I don't think I will like this.

Reply:
Let's go with duplicate codes. Sharing code each other will complicate dependency.

Reply:
https://github.com/Samsung/ONE/pull/13521/files#r1699149426